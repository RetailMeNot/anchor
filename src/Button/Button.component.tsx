// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled, { css, InterpolationValue } from 'styled-components';
import { darken } from 'polished';
// THEME
import { colors, fonts, sizes } from '../theme';

interface ButtonProps {
  children?: any;
  disabled?: boolean;
  showSpinner?: boolean;
  className?: string;

  onClick?: () => void;
  variant?: 'primary' | 'secondary' | 'text';
  size?: 'sm' | 'md' | 'lg';
}

interface ButtonStylesGroup {
  [key: string]: InterpolationValue[];
}

// TODO: a11y

const ButtonColorStyles: ButtonStylesGroup = {
  primary: css`
    border: solid thin ${colors.accent.savvyCyan.base};
    background-color: ${colors.accent.savvyCyan.base};
    color: ${colors.neutrals.white.base};

    &:hover {
      background-color: ${darken(0.05, colors.accent.savvyCyan.base)};
    }
  `,
  secondary: css`
    border: solid thin ${colors.accent.savvyCyan.light};
    background-color: ${colors.accent.savvyCyan.light};
    color: ${colors.neutrals.white.base};

    &:hover {
      background-color: ${darken(0.1, colors.accent.savvyCyan.light)};
    }
  `,
  text: css`
    border: solid thin ${colors.neutrals.white.base};
    background-color: ${colors.neutrals.white.base};
    color: ${colors.accent.savvyCyan.base};

    &:hover {
      color: ${darken(0.25, colors.accent.savvyCyan.base)};
    }
  `,
};

const ButtonSizeStyles: ButtonStylesGroup = {
  sm: css`
    padding: ${sizes.padding.xs};
    font-size: ${fonts.baseFontSize * 0.75}px;
  `,
  md: css`
    padding: ${sizes.padding.sm} ${sizes.padding.md};
    font-size: ${fonts.baseFontSize}px;
  `,
  lg: css`
    padding: ${sizes.padding.md};
    font-size: ${fonts.baseFontSize * 1.5}px;
  `,
};

export const StyledButton = styled.button`
	position: relative;
	min-width: 2rem;
	border-radius: ${sizes.border.radius.base};
	font-weight: 500;
	font-family: ${fonts.fontFamily};
	text-align: center;
	cursor: pointer;
  margin-right: ${sizes.padding.md};

	transition: 500ms ease background-color,
	            500ms ease border-color,
              500ms ease color,
              500ms ease fill;

	// These properties are deprecated but help make white text on colored backgrounds look more crisp in Chrome and
	// Firefox.
	-webkit-font-smoothing: antialiased;
	-moz-osx-font-smoothing: grayscale;

  /* Variants are color schemes */
  ${({ variant = 'primary' }: ButtonProps) => ButtonColorStyles[variant]}

  /* Sizes */
  ${({ size = 'md' }: ButtonProps) => ButtonSizeStyles[size]}

  /* Disabled State */
  ${({ disabled }: ButtonProps) => disabled && css`
    opacity: .5;
    cursor: not-allowed;
  `}

`;

const DefaultProps: ButtonProps = {};

export const Button = ({ className, ...props }: ButtonProps = DefaultProps): React.ReactElement<any> =>
  <StyledButton className={classNames(className)} {...props} />;

Button.defaultProps = DefaultProps;

export default Button;
