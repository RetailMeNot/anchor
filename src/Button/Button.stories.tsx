// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, boolean } from '@storybook/addon-knobs';
// import { checkA11y } from '@storybook/addon-a11y';
// VENDOR
import styled from 'styled-components';
// COMPONENT
import Button from './Button.component';

const StyledStory = styled.div`
  padding: 2rem 5rem;
`;

storiesOf('Components/Button', module)
  // .addDecorator(checkA11y) // TODO: wtf doesn't this work?
  .add('Primary', () => (
      <StyledStory>
          <Button
            disabled={boolean('Disabled', false)}
            onClick={action('onClick')}
          >
            {text('Button Label', 'Story Button')}
          </Button>
      </StyledStory>
  ))
  .add('Secondary', () => (
      <StyledStory>
          <Button
            disabled={boolean('Disabled', false)}
            variant="secondary"
          >{
            text('Button Label', 'Story Button')}
          </Button>
      </StyledStory>
  ))
  .add('Text', () => (
    <StyledStory>
      <Button
        variant="text"
      >
          {text('Button Label', 'Story Button')}
      </Button>
    </StyledStory>
  ));
