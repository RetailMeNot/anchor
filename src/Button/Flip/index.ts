export { Flip, StyledFlip } from './Flip.component';
