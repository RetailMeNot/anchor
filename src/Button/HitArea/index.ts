export { HitArea } from './HitArea.component';
