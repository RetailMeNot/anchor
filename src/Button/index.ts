export { default as Button } from './Button.component';
