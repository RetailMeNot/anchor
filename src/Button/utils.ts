export interface Theme {
    base: string;
    light: string;
    dark: string;
}

export const REVEAL_COLOR = '#F1F1F1';

export const TRANSITION_SPEED = '500ms'; // milliseconds
