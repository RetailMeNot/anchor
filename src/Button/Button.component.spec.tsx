// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// COMPONENT
import Button, { Button as Styles } from './Button.component';
// ENZYME
import { shallow, mount } from 'enzyme';

// TEST SETUP
const subject = <Button>Test</Button>;
const wrapper = mount(subject);
const component = shallow(subject);

describe('Component: Button', () => {
    it('should be defined', () => {
        /* tslint:disable */
        console.log(subject);
        console.log(wrapper.debug());
        console.log(component.debug());
        console.log(Styles);

        /* tslint:enable */
        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(<Button>Submit</Button>).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
