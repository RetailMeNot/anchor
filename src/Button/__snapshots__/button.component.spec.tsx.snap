// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Button should be defined 1`] = `
<button
  className="sc-bdVaJa gucETm"
>
  Submit
</button>
`;
